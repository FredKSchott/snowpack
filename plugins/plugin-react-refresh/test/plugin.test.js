mockBabel();
const path = require('path');
const fs = require('fs');
const pluginReactRefresh = require('../plugin');

const htmlFilePath = path.resolve(__dirname, './stubs/stub.html');
const htmlFileContent = fs.readFileSync(htmlFilePath, {
  encoding: 'utf-8',
});
const htmlTransformOptions = {
  contents: htmlFileContent,
  fileExt: '.html',
  id: 'stub.html',
  isDev: true,
  isHmrEnabled: true,
  isSSR: false,
};
const jsFilePath = path.resolve(__dirname, './stubs/stub.js');
const jsFileContent = fs.readFileSync(jsFilePath, {
  encoding: 'utf-8',
});
const jsTransformOptions = {
  contents: jsFileContent,
  fileExt: '.js',
  id: 'stub.js',
  isDev: true,
  isHmrEnabled: true,
  isSSR: false,
};

function mockBabel() {
  jest.mock('@babel/core');
  const babel = require('@babel/core');
  babel.transformAsync = jest
    .fn()
    .mockName('babel.transformAsync')
    .mockImplementation(async (contents, options, ...args) => {
      options.plugins = (options.plugins || []).map((plugin) => {
        if (Array.isArray(plugin)) {
          if (plugin[1]) {
            plugin[1].skipEnvCheck = true;
          }
          return plugin;
        }
        // Fix: React Refresh Babel transform should only be enabled in development environment.
        // Instead, the environment is: "test".
        // If you want to override this check, pass {skipEnvCheck: true} as plugin options.
        return [plugin, {skipEnvCheck: true}];
      });

      // Stop `jest.requireActual('@babel/core').transformAsync()` from requiring mocked babel function
      jest.unmock('@babel/core');
      const ret = await jest
        .requireActual('@babel/core')
        .transformAsync(contents, options, ...args);
      mockBabel();

      return ret;
    });
}

async function testPluginInstance(pluginInstance, overrides = {}) {
  const pluginTransform = pluginInstance.transform;
  expect(await pluginTransform({...htmlTransformOptions, ...overrides})).toMatchSnapshot('html');
  expect(await pluginTransform({...jsTransformOptions, ...overrides})).toMatchSnapshot('js');
}

afterEach(() => {
  jest.restoreAllMocks();
});

describe('@snowpack/plugin-react-refresh', () => {
  test('transform js and html', async () => {
    const pluginInstance = pluginReactRefresh({}, {babel: true});
    await testPluginInstance(pluginInstance);
  });
  test("don't transform when disabled", async () => {
    const pluginInstance = pluginReactRefresh({}, {babel: true});
    await testPluginInstance(pluginInstance, {isDev: false});
  });
  test('transform js and html when hmr is disabled', async () => {
    const pluginInstance = pluginReactRefresh({}, {babel: true});
    await testPluginInstance(pluginInstance, {isHmrEnabled: false});
  });

  test('transform js and html when running in SSR', async () => {
    const pluginInstance = pluginReactRefresh({}, {babel: true});
    await testPluginInstance(pluginInstance, {isSSR: true});
  });

  test('transform js and html when babel is disabled', async () => {
    const pluginInstance = pluginReactRefresh(
      {
        devOptions: {
          hmr: true,
        },
      },
      {babel: false},
    );
    await testPluginInstance(pluginInstance);
  });

  test('include custom babel config', async () => {
    const babel = require('@babel/core');
    const mockTransformAsync = jest.fn(() => Promise.resolve({code: ''}));
    jest.spyOn(babel, 'transformAsync').mockImplementation(mockTransformAsync);

    const babelConfig = {
      compact: true,
      plugins: ['@babel/plugin-syntax-top-level-await'],
    };
    const pluginInstance = pluginReactRefresh({}, {babel: babelConfig});
    await pluginInstance.transform({...jsTransformOptions});

    expect(mockTransformAsync).toHaveBeenCalledTimes(1);
    expect(mockTransformAsync.mock.calls[0][1]).toHaveProperty('compact', true);
    expect(mockTransformAsync.mock.calls[0][1].plugins).toContain(
      '@babel/plugin-syntax-top-level-await',
    );
  });
});
