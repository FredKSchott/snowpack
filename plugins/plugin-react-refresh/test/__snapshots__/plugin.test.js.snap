// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@snowpack/plugin-react-refresh hrm=false, babel=false, isDev=false transforms html correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=false, babel=false, isDev=false transforms js correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=false, babel=false, isDev=true transforms html correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=false, babel=false, isDev=true transforms js correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=false, babel=true, isDev=false transforms html correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=false, babel=true, isDev=false transforms js correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=false, babel=true, isDev=true transforms html correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=false, babel=true, isDev=true transforms js correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=true, babel=false, isDev=false transforms html correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=true, babel=false, isDev=false transforms js correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=true, babel=false, isDev=true transforms html correctly 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
<script src=\\"/node_modules/@snowpack/plugin-react-refresh/runtime.js\\"></script>
    <meta charset=\\"utf-8\\" />
    <link rel=\\"icon\\" href=\\"/favicon.ico\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />
    <meta name=\\"description\\" content=\\"Web site created using create-snowpack-app\\" />
  </head>
  <body>
    <div id=\\"root\\"></div>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <script type=\\"module\\" src=\\"/dist/index.js\\"></script>
  </body>
</html>"
`;

exports[`@snowpack/plugin-react-refresh hrm=true, babel=false, isDev=true transforms js correctly 1`] = `
"import React from 'react';
function App() {
  return /* @__PURE__ */ React.createElement('div', null, 'React App');
}
export default App;
"
`;

exports[`@snowpack/plugin-react-refresh hrm=true, babel=true, isDev=false transforms html correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=true, babel=true, isDev=false transforms js correctly 1`] = `undefined`;

exports[`@snowpack/plugin-react-refresh hrm=true, babel=true, isDev=true transforms html correctly 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
<script src=\\"/node_modules/@snowpack/plugin-react-refresh/runtime.js\\"></script>
    <meta charset=\\"utf-8\\" />
    <link rel=\\"icon\\" href=\\"/favicon.ico\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />
    <meta name=\\"description\\" content=\\"Web site created using create-snowpack-app\\" />
  </head>
  <body>
    <div id=\\"root\\"></div>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <script type=\\"module\\" src=\\"/dist/index.js\\"></script>
  </body>
</html>"
`;

exports[`@snowpack/plugin-react-refresh hrm=true, babel=true, isDev=true transforms js correctly 1`] = `
"
/** React Refresh: Setup **/
if (import.meta.hot) {
  if (!window.$RefreshReg$ || !window.$RefreshSig$ || !window.$RefreshRuntime$) {
    console.warn(\\"@snowpack/plugin-react-refresh: React Fast Refresh's runtime not found.\\\\nIt is served by default when Snowpack serves your HTML routes.\\\\nTo support other HTML files, add the following \`<script>\` tag immediately after /\`<head>\` tag, and before any other \`<script>\` tag:\\\\n\`<script src=\\\\\\"/node_modules/@snowpack/plugin-react-refresh/runtime.js\\\\\\"></script>\`\\");
  } else {
    var prevRefreshReg = window.$RefreshReg$;
    var prevRefreshSig = window.$RefreshSig$;
    window.$RefreshReg$ = (type, id) => {
      window.$RefreshRuntime$.register(type, \\"stub.js\\" + \\" \\" + id);
    }
    window.$RefreshSig$ = window.$RefreshRuntime$.createSignatureFunctionForTransform;
  }
}

import React from 'react';

function App() {
  return /* @__PURE__ */React.createElement('div', null, 'React App');
}

_c = App;
export default App;

var _c;

$RefreshReg$(_c, \\"App\\");

/** React Refresh: Connect **/
if (import.meta.hot) {
  window.$RefreshReg$ = prevRefreshReg
  window.$RefreshSig$ = prevRefreshSig
  import.meta.hot.accept(() => {
    window.$RefreshRuntime$.performReactRefresh()
  });
}"
`;
